### PR TITLE
[package] fix storage serialization collision

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61807
* #61805
* __->__ #61806

Currently, if you do `save_pickle` on a ScriptModule, then `save_pickle`
on a tensor, this would result in a `0.storage` tensor being written
*twice* to the zip archive. This would cause weird bugs on the
serializing side (this presented as a ASAN-detected heap buffer overflow
because we tried to read more memory from a tensor than we actually
had).

Turns out this was because when we did:
```
self.storage_context = self.script_module_serializer.storage_context()
```
it returned a new copy of the storage context, so we weren't actually
assigning unique names to tensors!!

This PR fixes the issue by making `(De)SerializationStorageContext`
non-copyable and fixing up the parts of the bindings that returned by
copy.

Differential Revision: [D29748969](https://our.internmc.facebook.com/intern/diff/D29748969)

### CODE REVIEW

**Comment by suo** [comment]
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29748969).

**Comment by suo** [comment]
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29748969).

File: test/package/test_torchscript.py [hunk 1]
```diff
@@ -486,6 +486,36 @@ def test_load_shared_tensors_repackaged(self):
...
+        object causes no issues.
+        """
+
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.foo = torch.ones(2, 3)
+
+            def forward(self):
+                return self.foo
```
Comment (Lilyjjo) [inline test/package/test_torchscript.py:500]: You could use 'ModWithTensor' from package_a.test_module if you wanted to and I think it would fill the need

**Review by lobstergrindset — APPROVED** [review]
LGTM! Thanks for figuring this out!

**Comment by suo** [comment]
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29748969).

**Pull request closed without merging**